File: sdks/c++/basys.sdk.cc/src/submodel/submodel/api/dataspecification/datatypes/DataTypeIEC61360.cpp
#include <BaSyx/submodel/api/dataspecification/datatypes/DataTypeIEC61360.h>

#include <unordered_map>
#include <string>

#include <BaSyx/util/util.h>

using namespace basyx::submodel;

static const std::string DataTypeIEC61360_to_string[] =
{
	"Date",
	"String",
	"String_Translatable",
	"Real_Measure",
	"Real_Count",
	"Real_Currency",
	"Boolean",
	"Url",
	"Rational",
	"Rational_Measure",
	"Time",
	"Timestamp"
};

static const std::unordered_map<std::string, basyx::submodel::DataTypeIEC61360> string_to_DataTypeIEC61360
{
	{DataTypeIEC61360_to_string[static_cast<char>(DataTypeIEC61360::Date)], DataTypeIEC61360::Date},
	{DataTypeIEC61360_to_string[static_cast<char>(DataTypeIEC61360::String)], DataTypeIEC61360::String},
	{DataTypeIEC61360_to_string[static_cast<char>(DataTypeIEC61360::String_Translatable)], DataTypeIEC61360::String_Translatable},
	{DataTypeIEC61360_to_string[static_cast<char>(DataTypeIEC61360::Real_Measure)], DataTypeIEC61360::Real_Measure},
	{DataTypeIEC61360_to_string[static_cast<char>(DataTypeIEC61360::Real_Count)], DataTypeIEC61360::Real_Count},
	{DataTypeIEC61360_to_string[static_cast<char>(DataTypeIEC61360::Real_Currency)], DataTypeIEC61360::Real_Currency},
	{DataTypeIEC61360_to_string[static_cast<char>(DataTypeIEC61360::Boolean)], DataTypeIEC61360::Boolean},
	{DataTypeIEC61360_to_string[static_cast<char>(DataTypeIEC61360::Url)], DataTypeIEC61360::Url},
	{DataTypeIEC61360_to_string[static_cast<char>(DataTypeIEC61360::Rational)], DataTypeIEC61360::Rational},
	{DataTypeIEC61360_to_string[static_cast<char>(DataTypeIEC61360::Rational_Measure)], DataTypeIEC61360::Rational_Measure},
	{DataTypeIEC61360_to_string[static_cast<char>(DataTypeIEC61360::Time)], DataTypeIEC61360::Time},
	{DataTypeIEC61360_to_string[static_cast<char>(DataTypeIEC61360::Timestamp)], DataTypeIEC61360::Timestamp}
};

const std::string & util::to_string(basyx::submodel::DataTypeIEC61360 type)
{
	return DataTypeIEC61360_to_string[static_cast<char>(type)];
}

template<>
basyx::submodel::DataTypeIEC61360 util::from_string<basyx::submodel::DataTypeIEC61360>(const std::string & str)
{
	return string_to_DataTypeIEC61360.at(str);
}

File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/submodel/metamodel/api/dataspecification/enums/DataTypeIEC61360.java
package org.eclipse.basyx.submodel.metamodel.api.dataspecification.enums;

import org.eclipse.basyx.submodel.metamodel.enumhelper.StandardizedLiteralEnum;
import org.eclipse.basyx.submodel.metamodel.enumhelper.StandardizedLiteralEnumHelper;

/**
 * Possible value data types as defined in DAAS for IEC61360 data specification templates
 * 
 * @author espen
 *
 */
public enum DataTypeIEC61360 implements StandardizedLiteralEnum {
	BOOLEAN("BOOLEAN"), DATE("DATE"), RATIONAL("RATIONAL"), RATIONAL_MEASURE("RATIONAL_MEASURE"),
	REAL_COUNT("REAL_COUNT"), REAL_CURRENCY("REAL_CURRENCY"), REAL_MEASURE("REAL_MEASURE"), STRING("STRING"),
	STRING_TRANSLATABLE("STRING_TRANSLATABLE"),
	TIME("TIME"), TIME_STAMP("TIME_STAMP"), URL("URL");

	private String standardizedLiteral;

	private DataTypeIEC61360(String standardizedLiteral) {
		this.standardizedLiteral = standardizedLiteral;
	}

	@Override
	public String getStandardizedLiteral() {
		return standardizedLiteral;
	}

	@Override
	public String toString() {
		return standardizedLiteral;
	}

	public static DataTypeIEC61360 fromString(String str) {
		return StandardizedLiteralEnumHelper.fromLiteral(DataTypeIEC61360.class, str);
	}
}
